File: Python/Algorithm/Algorithm I/704. Binary Search.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        """
        Search for the target value in a sorted list of integers using binary search.

        Args:
            nums (List[int]): A sorted list of integers.
            target (int): The target value to search for.
        
        Returns:
            int: The index of the target value if found, otherwise -1.

        Time Complexity: o(log(n)) where n is the length of the list.
        Space Complexity: o(1) since we only use three integer variables.
        """
        NUMS_LENGTH = len(nums)
        
        leftPointer = 0
        rightPointer = NUMS_LENGTH - 1

        while leftPointer <= rightPointer:
            middlePointer = leftPointer + (rightPointer - leftPointer) // 2
            middleValue = nums[middlePointer]

            if middleValue == target:
                return middlePointer

            elif middleValue < target:
                leftPointer = middlePointer + 1

            else:
                rightPointer = middlePointer - 1

        return -1

File: Java/Algorithm/Algorithm I/704. Binary Search.java
class Solution {

  /**
   * Searches for the target element in a sorted array using binary search.
   *
   * @param nums   The sorted array to search in.
   * @param target The target element to find.
   * @return The index of the target element if found, otherwise -1.
   */
  public int search(int[] nums, int target) {
    int NUMS_LENGTH = nums.length;

    int leftPointer = 0;
    int rightPointer = NUMS_LENGTH - 1;

    while (leftPointer <= rightPointer) {
      int middlePointer = leftPointer + (rightPointer - leftPointer) / 2;
      int middleValue = nums[middlePointer];

      if (middleValue == target) 
        return middlePointer; 
        
      else if (middleValue > target) 
        rightPointer = middlePointer - 1; 
        
      else 
        leftPointer = middlePointer + 1;
    }

    return -1;
  }
}
